# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

I belive that random forest will perform better as it is better for larger decision trees and logistic regression is more fit for more simple models

## Split the Data into Training and Testing Sets

In [4]:
# Create the labels set `y` and features DataFrame `X`
y = data["spam"]
X = data.drop("spam", axis=1)



In [20]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()


spam
0    2788
1    1813
Name: count, dtype: int64

In [6]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [7]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
s_scaler = StandardScaler()


In [8]:
# Fit the Standard Scaler with the training data
s_scaler.fit(X_train)


StandardScaler()

In [12]:
# Scale the training data
X_train_scaled = s_scaler.transform(X_train)

X_train_scaled

# Scale the testing data
X_test_scaled = s_scaler.transform(X_test)

X_test_scaled


array([[-0.32942636, -0.16291587, -0.55249434, ..., -0.08787279,
        -0.24211005, -0.35762644],
       [-0.32942636, -0.16291587, -0.55249434, ..., -0.02380129,
        -0.2745267 , -0.3689973 ],
       [-0.32942636, -0.16291587, -0.55249434, ..., -0.08946289,
        -0.32315167, -0.33975795],
       ...,
       [-0.32942636,  0.86660069, -0.55249434, ..., -0.09482556,
        -0.20969341, -0.28452807],
       [-0.32942636, -0.16291587, -0.55249434, ..., -0.07674212,
        -0.29883918, -0.32188946],
       [ 0.14679617, -0.04226939,  0.06652667, ...,  0.0417356 ,
         1.20043064,  0.53742255]])

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [16]:
# Train a Logistic Regression model
from sklearn.linear_model import LogisticRegression

regression_model = LogisticRegression()
regression_model.fit(X_train_scaled, y_train)

# Print the model score
print(regression_model.score(X_test_scaled, y_test))


0.9252823631624674


In [13]:
# Make and save testing predictions with the saved logistic regression model using the test data
test_predictions = regression_model.predict(X_test_scaled)

# Review the predictions
test_predictions

array([1, 0, 0, ..., 0, 0, 1], dtype=int64)

In [15]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy = accuracy_score(y_test, test_predictions)
accuracy

0.9252823631624674

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [17]:
# Train a Random Forest Classifier model
from sklearn.ensemble import RandomForestClassifier

forest_model = RandomForestClassifier()
forest_model.fit(X_train_scaled, y_train)

# Print the model score
print(forest_model.score(X_test_scaled, y_test))


0.94874022589053


In [18]:
# Make and save testing predictions with the saved logistic regression model using the test data
test_predictions = forest_model.predict(X_test_scaled)

# Review the predictions
test_predictions

array([1, 0, 0, ..., 0, 0, 1], dtype=int64)

In [19]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy = accuracy_score(y_test, test_predictions)
accuracy

0.94874022589053

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

The random forest model performed better with an accuracy score of 95% vs the linear regression model which had an accuracy score of 93%. These results prove my earlier prediction correct.